crates/transaction-pool/src/maintain.rs

```diff
@@ -0,0 +1,268 @@
+                    continue
+                }
+
+                // base fee for the next block: `first_block+1`
+                let pending_block_base_fee = calculate_next_block_base_fee(
+                    first_block.gas_used,
+                    first_block.gas_limit,
+                    first_block.base_fee_per_gas.unwrap_or_default(),
+                ) as u128;
+                let changed_accounts = changed_accounts_iter(state).collect();
+                let update = CanonicalStateUpdate {
+                    hash: first_block.hash,
+                    number: first_block.number,
+                    pending_block_base_fee,
+                    changed_accounts,
+                    // no tx to prune in the reverted chain
+                    mined_transactions: vec![],
+                };
+                pool.on_canonical_state_change(update);
+
+                let pruned_old_transactions = blocks
+                    .transactions()
+                    .filter_map(|tx| tx.clone().into_ecrecovered())
+                    .map(|tx| {
+                        <V as TransactionValidator>::Transaction::from_recovered_transaction(tx)
+                    })
+                    .collect();
+
+                // all transactions that were mined in the old chain need to be re-injected
+                //
+                // Note: we no longer know if the tx was local or external
+                let _ = pool.add_external_transactions(pruned_old_transactions).await;
+                // TODO: metrics
+            }
+            CanonStateNotification::Commit { new } => {
+                // TODO skip large commits?
+
+                let (blocks, state) = new.inner();
+                let tip = blocks.tip();
+                // base fee for the next block: `tip+1`
+                let pending_block_base_fee = calculate_next_block_base_fee(
+                    tip.gas_used,
+                    tip.gas_limit,
+                    tip.base_fee_per_gas.unwrap_or_default(),
+                ) as u128;
+
+                let first_block = blocks.first();
+                // check if the range of the commit is canonical
+                if first_block.parent_hash == pool_info.last_seen_block_hash {
+                    let changed_accounts = changed_accounts_iter(state).collect();
+                    let mined_transactions = blocks.transactions().map(|tx| tx.hash).collect();
+                    // Canonical update
+                    let update = CanonicalStateUpdate {
+                        hash: tip.hash,
+                        number: tip.number,
+                        pending_block_base_fee,
+                        changed_accounts,
+                        mined_transactions,
+                    };
+                    pool.on_canonical_state_change(update);
+                } else {
+                    // TODO is this even reachable, because all commits are canonical?
+                    // this a canonical
```

Comment:
should be reachable in case of `tree -> pipeline -> tree`, so pool_info ends up in bad state, right?

Reply:
commits are all canonical,right? so I'm still skeptical

Reply:
commits in the blockchain tree are. but the pipeline doesn't emit these events, right? so, the loose `first_block.parent_hash == pool_info.last_seen_block_hash` condition would be false because pool info hasn't been updated with information that pipeline inserted

Reply:
> commits in the blockchain tree are. but the pipeline doesn't emit these events, right? 

not yet, but we need this, needs be emitted when we inserted a block.
will open an issue for this.